api/src/services/users.js

```diff
@@ -432,51 +480,218 @@ const missingFields = data => {
+ */
+const manageTokenLogin = (data, response) => {
+  if (shouldDisableTokenLogin(data)) {
+    return disableTokenLogin(response);
+  }
+
+  if (!shouldEnableTokenLogin(data)) {
+    return Promise.resolve(response);
+  }
+
+  return enableTokenLogin(response);
+};
+
+/**
+ * Enables token-login for a user
+ * - a new document is created in the `medic` database that contains tasks (SMSs) to be sent to the phone number
+ *   belonging to the current user (userSettings.phone) that contain instructions and a url to login in the app
+ * - if the user already had token-login enabled, the previous sms document's tasks are cleared
+ * - updates the user document to contain information about the active `token-login` context for the user
+ * - updates the user-settings document to contain some information to be displayed in the admin page
+ * @param {Object} response - the response of previous actions
+ * @returns {Promise<{Object}>} - updated response to be sent to the client
+ */
+const enableTokenLogin = (response) => {
+  return Promise
+    .all([
+      validateUser(response.user.id),
+      validateUserSettings(response['user-settings'].id),
+    ])
+    .then(([ user, userSettings ]) => {
+      const tokenLoginConfig = config.get('token_login');
+      const token = generatePassword(50);
+
+      return generateLoginSms(user, userSettings, token, tokenLoginConfig)
+        .then(result => {
+          user.token_login = {
+            active: true,
+            token,
+            expiration_date: new Date().getTime() + LOGIN_TOKEN_EXPIRE_TIME,
+            doc_id: result.id
+          };
+
+          userSettings.token_login = {
+            active: true,
+            expiration_date: user.token_login.expiration_date,
+          };
+
+          response.token_login = {
+            id: result.id,
+            expiration_date: user.token_login.expiration_date,
+          };
+
+          return Promise.all([ db.users.put(user), db.medic.put(userSettings) ]);
+        });
+    })
+    .then(() => response);
+};
+
+/**
+ * Disables token-login for a user.
+ * Deletes the `token_login` properties from the user and userSettings doc.
+ * Clears pending tasks in existent SMSs
+ *
+ * @param {Object} response - the response of previous actions
+ * @returns {Promise<{Object}>} - updated response to be sent to the client
+ */
+const disableTokenLogin = (response) => {
+  return Promise
+    .all([
+      validateUser(response.user.id),
+      validateUserSettings(response['user-settings'].id),
+    ])
+    .then(([ user, userSettings ]) => {
+      if (!user.token_login) {
+        return;
+      }
+
+      return clearOldLoginSms(user).then(() => {
+        delete user.token_login;
+        delete userSettings.token_login;
+
+        return Promise.all([
+          db.medic.put(userSettings),
+          db.users.put(user),
+        ]);
+      });
+    })
+    .then(() => response);
+};
```

Comment:
This service is getting huge! I think all the login token handling code should be moved into a new service to make it a more reasonable size. This new service could also handle all the token login code reducing the size of the login controller?

Reply:
Yea, I agree. I'll shuffle things around!